File: src/leetcode/python/ShortestUnsortedContinuousSubarray_581.py
class ShortestUnsortedContinuousSubarray_581(object):
    def findUnsortedSubarray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        end, start = -2, -1
        ma, mi = nums[0], nums[-1]
        n = len(nums)
        for i in range(1, n):
            ma = max(nums[i], ma)
            mi = min(nums[n - i - 1], mi)
            if nums[i] < ma:
                end = i
            if nums[n - i - 1] > mi:
                start = n - i - 1
        return end - start + 1

File: src/leetcode/java/ShortestUnsortedContinuousSubarray_581.java
package leetcode.java;

public class ShortestUnsortedContinuousSubarray_581 {
  public int findUnsortedSubarray(int[] nums) {
    int end = -2, start = -1;
    int n = nums.length;
    int min = nums[n - 1], max = nums[0];
    for (int i = 1; i < n; i++) {
      max = Math.max(nums[i], max);
      min = Math.min(nums[n - i - 1], min);
      if (nums[i] < max) end = i;
      if (nums[n - i - 1] > min) start = n - i - 1;
    }
    return end - start + 1;
  }
}
